### ISSUE TITLE
Did moving sounddevice.py break _sounddevice_data installation?

### ISSUE THREAD

**Issue by tgarc** [spatialaudio/python-sounddevice#110]
So in commit 6e1d71547 you moved sounddevice.py under src and I'm just wondering if this doesn't break the libportaudio load that is attempted at the top of sounddevice.py. If sounddevice.py still gets installed directly to the site-packages directory (i.e. not in a subdirectory) than it shouldn't be an issue but I can't tell if this is the behavior on all platforms. Can't test this on windows/macos at the moment hence why I'm asking.

BTW this came up because I had sounddevice installed in editable mode (pip install -e <local-sounddevice-directory>) and with this setup _sounddevice_data is actually up a directory from sounddevice.py; so in that case the libportaudio load is does not work properly.

**Comment by mgeier** [comment]
Yes, sorry, I missed that.

I think it should be OK when simply installing it, although I'm not 100% sure.

But the logic for loading the shared library is indeed broken for editable installs.

A more reliable way to find the path could be to import the `_sounddevice_data` package:

```python
import _sounddevice_data
_lib = _ffi.dlopen(_os.path.join(next(iter(_sounddevice_data.__path__)), _libname))
```

Sadly, this doesn't work in Python 2, because we don't have a file `_sounddevice_data/__init__.py`.

We could add such a `__init__.py`, but this would probably look a bit strange in https://github.com/spatialaudio/portaudio-binaries.

Alternatively, we could just check if the path to `__file__` ends with `src`.

What do you think? Do you have a better idea?

**Comment by tgarc** [comment]
Not sure what would be a better approach atm, just wanted to bring it up.
I'm guessing you wouldn't want to install libportaudio and sounddevice in
the same directory since (I believe) that would cause ffi to load *it*
instead of any possible system installed portaudio.

As long as the current approach does indeed work for regular installs then
I'm in no hurry to find a fix for this.

**Comment by mgeier** [comment]
> I'm guessing you wouldn't want to install libportaudio and sounddevice in the same directory since

You are right, I wouldn't want that.
Even if it wouldn't interfere with loading the right version of PortAudio, I think it's much cleaner to have the additional files in a sub-directory and not clutter the main installation directory.
I think it's nicer if we only install files whose paths begin with `sounddevice` and `_sounddevice`.

I've updated #112, can you please check if that works for you for an "editable" installation?

**Comment by tgarc** [comment]
#112 works great. Maybe having an __init__.py is weird but your solution is robust and even works with editable install

**Comment by tgarc** [comment]
Also, this patch should probably be applied to pysoundfile as well for the same reasons

**Comment by mgeier** [comment]
@tgarc Thanks for testing!

The `soundfile` module has a special hack for cx_Freeze, but I guess this could also be adapted to using `import`.

**Comment by tgarc** [comment]
so it looks like I may have lied. This in fact doesn't work with editable installs. When I first tested it I was running it from the root of the repository so ``import _sounddevice_data`` was just picking up on the directory from the repository. Anyway, no big deal but just wanted to correct my statement.

**Comment by mgeier** [comment]
Hmmm, this is unfortunate.

I found two rather recent questions with no real answers:

https://stackoverflow.com/questions/45139571/install-package-with-separate-source-directory-in-editable-mode
https://github.com/pypa/pip/issues/3160

Then I found an older comment explaining the problem:

https://github.com/pypa/pip/issues/126#issuecomment-874855

... which links to a `distribute` issue:

https://bitbucket.org/tarek/distribute/issues/177/setuppy-develop-

... which links (after correcting the dead URL) to:

https://github.com/pypa/setuptools/issues/230

... which is still open and there is some very recent discussion about a way how to solve this problem:

https://mail.python.org/pipermail/distutils-sig/2017-September/031498.html

All in all, this doesn't look like it will be resolved anytime soon.

I guess I'll have to get rid of `src/` and move `sounddevice.py` back to the main directory.

Or do you have another idea?

**Comment by tgarc** [comment]
Far as I can tell, editable installs are just a convenience; I wouldn't
move back sounddevice.py just to make that work. Plus, editable installs
still work as long as you run it from the root directory.

OTOH, I suspect that moving _sounddevice_data under src/ would fix the
issue...but then you end up with binaries in a folder called src.

**Comment by mgeier** [comment]
I thought about moving `_sounddevice_data` to `src/`, but this would completely contradict the idea of moving *source* files to a separate place. One of the ideas of the `src/` directory was to make it easy to locate the files where the actual code happens.

And yes, editable installs are "just" a convenience, but I would like to make contributions as easy as possible, even for Windows users (macOS users can rather easily install PortAudio on their system to avoid the current issue, but it's still some work).

Both ways have minor advantages and disadvantages, but I think I prefer to have more convenience for contributors over a somewhat cleaner directory layout.

**Comment by mgeier** [comment]
I've created #113 to get rid of `src/`.
If there are any objections, I'm of course still open!

**Comment by mgeier** [comment]
I've merged #113, `src/` is gone.

If the PyPA fixes the problems in `setuptools` at some point, we can re-introduce it.
Or we find another nice solution some day, who knows ...

**Comment by sprocket1** [comment]
Hi, I'm very new to Python and Spyder.
I'm getting this on the Spyder console: 

import sounddevice as sd
Traceback (most recent call last):

  File "<ipython-input-1-1b8d9fe8fdfa>", line 1, in <module>
    import sounddevice as sd

  File "C:\Users\James\Anaconda3\lib\site-packages\sounddevice.py", line 73, in <module>
    import _sounddevice_data

ModuleNotFoundError: No module named '_sounddevice_data'

Not sure what's going on, but seems related to this thread.
